src/plugins/jingle/utils.js

```diff
@@ -0,0 +1,119 @@
+import { converse, api, _converse } from '@converse/headless/core';
+import JingleCallModal from "./modal/jingle-incoming-call-modal.js";
+import tpl_jingle_chat_history from "./templates/jingle_chat_history.js";
+
+const { Strophe, sizzle, $msg } = converse.env;
+const u = converse.env.utils;
+
+
+/**
+ * This function merges the incoming attributes and the jingle propose attribute into one.
+ * It also determines the type of the media i.e. audio or video
+ * @param { XMLElement } stanza
+ * @param { Object } attrs
+ */
+export function parseJingleMessage(stanza, attrs) {
+    const jingle_propose_type = getJingleProposeType(stanza);
+    // To do editable: false, retracted_id: {if there is a retractions set it to the id}, retracted(timestamp)
+    return { ...attrs, ...{ 'jingle_propose': jingle_propose_type, 'jingle_retraction_id': getJingleRetractionID(stanza), 'template_hook': (attrs['template_hook']) ? 'getJingleTemplate' : undefined, 'jingle_status': attrs['jingle_status'] }}
+}
+
+function getJingleProposeType(stanza){
+    const el = sizzle(`propose[xmlns="${Strophe.NS.JINGLEMESSAGE}"] > description`, stanza).pop();
+    return el?.getAttribute('media');
+}
+
+function getJingleRetractionID(stanza){
+    const el = sizzle(`propose[xmlns="${Strophe.NS.JINGLEMESSAGE}"]`, stanza).pop();
+    return el?.getAttribute('id');
+}
+
+export function getJingleTemplate(model) {
+    return tpl_jingle_chat_history(model);
+}
+
+export function jingleCallInitialized() {
+    JingleCallModal;
+}
+
+/**
+ * This function simply sends the retraction stanza and modifies the attributes
+ */
+export function retractCall(context) {
+    const initiator_message = context.model.messages.findWhere({ 'media': 'audio' });
```

Comment:
@jcbrand I believe findWhere isn't working, I tested it several times, and yeah, if we use where, I am getting an array of 150 + to propose messages, which shouldn't happen I think?